### PR TITLE
Adaptive compression training loop improvements

### PR DESCRIPTION
**Pull request by nikita-savelyevv**
This PR finishes up the changes started in #1134.

### Changes
- Making use of EarlyExit loop as initial phase of AdaptiveCompression loop. This reduces the time required for initial training as it may now exit early.
- Introduction early exit -like logic to main AdaptiveCompression loop in the form of existing a training phase if positive accuracy budget is reached. This further reduces the training time.
- Minor changes:
  - Improved logging
  - Code inspection warnings fixes
  - Reduced the number of parameters which user needs to set manually
- Bugs fixed:
  - Failing updating initial scheduler learning rate ['base_lrs', 'init_lr’]
  - Compression scheduler disabling behavior was incorrect
  - Incorrect behavior of best model saving for sparsity backend

### Reason for changes

Improvement and bug fixing of AdaptiveCompression algorithm

### Related tickets

94288

### Tests

Existing tests were tweaked according to the new algorithm logic. New tests addition is left for the future ticket. 

Performed experiments on ResNet18 CIFAR10 for pruning backend. Experiments demonstrate improvement of training time without significant loss of accuracy or compression.
| Value | Before | This PR |
| ----------- | ----------- | ----- |
| Epochs trained | 350 |   137  |
| Absolute accuracy drop | 0.74 % |  0.84 %  |
| Compression rate | 47.5 %   |  47.5 %  |

Slighter lower accuracy drop for old algorithm may be due to overfitting to the validation dataset.

Further TODOs / self-review is in the [separate branch](https://github.com/nikita-savelyevv/nncf/tree/actl-todo).

### CODE REVIEW

File: nncf/common/accuracy_aware_training/runner.py [hunk 1]
```diff
@@ -20,16 +21,47 @@
 from typing import TypeVar
 from abc import ABC
 from abc import abstractmethod
+import io
+import os
+import os.path as osp
```
Comment (daniil-lyakhov) [inline nncf/common/accuracy_aware_training/runner.py:26]: Please move this imports to the top of the file to meet [NNCF styleguide](https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md) reqirements

File: examples/torch/classification/main.py [hunk 2]
```diff
@@ -259,7 +259,7 @@ def validate_fn(model, epoch):
             # training function that trains the model for one epoch (full training dataset pass)
             # it is assumed that all the NNCF-related methods are properly called inside of
             # this function (like e.g. the step and epoch_step methods of the compression scheduler)
-            def train_epoch_fn(compression_ctrl, model, epoch, optimizer, **kwargs):
+            def train_epoch_fn(compression_ctrl, model, optimizer, lr_scheduler, epoch):
```
Comment (daniil-lyakhov) [inline examples/torch/classification/main.py:262]: Does this change needed in in segmentation/detection samples?

Comment (nikita-savelyevv) [reply examples/torch/classification/main.py:262]: Not critical, but better to change indeed

File: nncf/common/accuracy_aware_training/runner.py [hunk 3]
```diff
@@ -277,42 +401,134 @@ def initialize_training_loop_fns(self, train_epoch_fn: Callable[[CompressionAlgo
...
-        self._log_dir = log_dir
 
-    def calculate_minimal_tolerable_accuracy(self, uncompressed_model_accuracy: float):
-        if self.maximal_absolute_accuracy_drop is not None:
-            self.minimal_tolerable_accuracy = uncompressed_model_accuracy - self.maximal_absolute_accuracy_drop
-        else:
-            self.minimal_tolerable_accuracy = uncompressed_model_accuracy * \
-                                              (1 - 0.01 * self.maximal_relative_accuracy_drop)
+        self._log_dir = log_dir if log_dir is not None else osp.join(os.getcwd(), 'runs')
+        self._log_dir = configure_accuracy_aware_paths(self._log_dir)
```
Comment (kshpv) [inline nncf/common/accuracy_aware_training/runner.py:406]: You should call this only in main process, because it could lead to problems in distributed mode when several process create log_fir independetly. Take a look at https://github.com/openvinotoolkit/nncf/pull/1371

**Review by kshpv — CHANGES_REQUESTED** [review]

**Comment by nikita-savelyevv** [comment]
TF precommit failed due to CI issues. [Rebuild ](https://ci-adas-icv.iotg.sclab.intel.com/job/NNCF/job/precommit/job/ubuntu20_precommit_tensorflow/829/) is successful.

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 4]
```diff
@@ -53,38 +54,28 @@ def run(self, model: TModel, train_epoch_fn, validate_fn, configure_optimizers_f
...
-        self.runner = runner_factory.create_training_loop()
-        self.compression_controller = compression_controller
+class BaseEarlyExitCompressionTrainingLoop(TrainingLoop, ABC):
+    def __init__(self):
+        self.runner = None
+        self.compression_controller = None
 
-    def run(self, model, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
-            dump_checkpoint_fn=None, tensorboard_writer=None, log_dir=None):
+    def run(self, model: TModel, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/training_loop.py:71]: This could use more type hints

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 5]
```diff
@@ -173,7 +169,30 @@ def _accuracy_criterion_satisfied(accuracy_budget, compression_controller):
...
+    Adaptive compression training loop allows an accuracy-aware training process
+    to reach the maximal accuracy drop
+    (the maximal allowed accuracy degradation criterion is satisfied).
+    """
+
+    def __init__(self,
+                 nncf_config: NNCFConfig,
+                 compression_controller: CompressionAlgorithmController,
+                 lr_updates_needed=True, verbose=True,
+                 dump_checkpoints=True):
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/training_loop.py:183]: Set `bool` type hints here as well.

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 6]
```diff
@@ -230,52 +251,78 @@ def remove_registry_prefix(algo_name):
         raise RuntimeError('No compression algorithm that supports adaptive compression '
                            'accuracy-aware training was specified')
 
-    def run(self, model, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
-            dump_checkpoint_fn=None, tensorboard_writer=None, log_dir=None):
+    def run(self, model: TModel, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/training_loop.py:254]: Type hints

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 7]
```diff
@@ -230,52 +251,78 @@ def remove_registry_prefix(algo_name):
...
+                nncf_logger.info('The target compression rate value was changed {prev_comp_rate:.3f} -> '
+                                 '{comp_rate:.3f}'.format(prev_comp_rate=prev_compression_rate_target,
+                                                          comp_rate=self.runner.compression_rate_target))
+                if prev_compression_rate_step == self.runner.compression_rate_step:
+                    nncf_logger.info('The compression rate step value was not changed: '
+                                     '{comp_step:.3f}'.format(comp_step=self.runner.compression_rate_step))
+                else:
+                    nncf_logger.info('The compression rate step value was changed {prev_comp_step:.3f} -> '
+                                     '{comp_step:.3f}'.format(prev_comp_step=prev_compression_rate_step,
+                                                              comp_step=self.runner.compression_rate_step))
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/training_loop.py:306]: Consider refactoring these to use f-strings for brevity

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 8]
```diff
@@ -284,67 +331,79 @@ def run(self, model, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
...
-            runner.train_epoch(model, accuracy_aware_controller)
-        compressed_model_accuracy = runner.validate(model)
-        runner.accuracy_bugdet = compressed_model_accuracy - runner.minimal_tolerable_accuracy
-        runner.add_tensorboard_scalar('val/accuracy_aware/accuracy_bugdet',
-                                      runner.accuracy_bugdet, runner.cumulative_epoch_count)
-        nncf_logger.info('Accuracy budget value after training is {}'.format(runner.accuracy_bugdet))
-
-    def _update_target_compression_rate(self, accuracy_aware_controller, runner):
-        current_compression_rate = accuracy_aware_controller.compression_rate
+    def _run_initial_training_phase(self, model):
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/training_loop.py:363]: Type hints

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 9]
```diff
@@ -284,67 +331,79 @@ def run(self, model, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
...
+
+        maximal_total_epochs = self.runner.maximal_total_epochs
+        self.runner.maximal_total_epochs = self.runner.initial_training_phase_epochs
+        model = self._run_early_exit_training_loop(model)
+        self.runner.maximal_total_epochs = maximal_total_epochs
+
+        nncf_logger.info('Finish the initial learning phase')
+        return model
+
+    def _update_target_compression_rate(self, runner, force_update=False):
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/training_loop.py:374]: Type hints everywhere

File: nncf/tensorflow/accuracy_aware_training/runner.py [hunk 10]
```diff
@@ -13,132 +13,104 @@
...
 
 
 class TFAccuracyAwareTrainingRunner(BaseAccuracyAwareTrainingRunner):
     """
     The Training Runner implementation for TensorFlow training code.
     """
 
-    def initialize_training_loop_fns(self, train_epoch_fn, validate_fn, configure_optimizers_fn=None,
-                                     dump_checkpoint_fn=None, tensorboard_writer=None, log_dir=None):
+    checkpoint_path_extension = '.pt'
```
Comment (vshampor) [inline nncf/tensorflow/accuracy_aware_training/runner.py:31]: Make constant names all-caps

File: nncf/tensorflow/accuracy_aware_training/runner.py [hunk 11]
```diff
@@ -13,132 +13,104 @@
...
+        is_best_checkpoint = (self.best_val_metric_value == self.current_val_metric_value and
+                              isinstance(compression_controller.scheduler, StubCompressionScheduler))
+        if not self.dump_checkpoints and not is_best_checkpoint:
+            return
+
+        if self._dump_checkpoint_fn is not None:
+            checkpoint_path = self._dump_checkpoint_fn(model, compression_controller, self, self._checkpoint_save_dir)
+        else:
+            checkpoint_path = osp.join(self._checkpoint_save_dir,
+                                       f'acc_aware_checkpoint_last{self.checkpoint_path_extension}')
```
Comment (vshampor) [inline nncf/tensorflow/accuracy_aware_training/runner.py:111]: Refactor to use `pathlib`

File: examples/torch/classification/main.py [hunk 12]
```diff
@@ -266,7 +266,7 @@ def validate_fn(model, epoch):
...
```
Comment (alexsu52) [inline examples/torch/classification/main.py:269]: Why did you change signature of the `train_epoch_fn` function? 

The `lr_scheduler` is not used in the `train_epoch_fn` function. It looks like a bad style.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/api/compression.py [hunk 13]
```diff
@@ -13,11 +13,11 @@
 
 from abc import ABC
 from abc import abstractmethod
+from enum import IntEnum
```
Comment (alexsu52) [inline nncf/api/compression.py:16]: We already merged it. Please, rebase to develop.

File: nncf/common/accuracy_aware_training/runner.py [hunk 14]
```diff
@@ -262,6 +313,79 @@ def __init__(self, accuracy_aware_params: Dict[str, object], verbose=True,
...
+        self._train_epoch_fn = None
+        self._validate_fn = None
+        self._configure_optimizers_fn = None
+        self._dump_checkpoint_fn = None
+
+        self._log_dir = None
+        self._checkpoint_save_dir = None
+        self._tensorboard_writer = None
+
+    def train_epoch(self, model, compression_controller):
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/runner.py:336]: Set missing type hins everywhere

File: examples/tensorflow/classification/main.py [hunk 15]
```diff
@@ -256,8 +255,6 @@ def run(config):
                                               callbacks=callbacks,
                                               initial_epoch=initial_epoch,
                                               steps_per_epoch=train_steps,
-                                              tensorboard_writer=SummaryWriter(config.log_dir,
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply examples/tensorflow/classification/main.py:259]: Runners require a raw tensorboard writer because it calls `.add_scalar()` and `.add_image()` methods. This `SummaryWriter` is our custom summary writer and it does not implement these methods

File: examples/torch/classification/main.py (hunk 12 above)
Comment (nikita-savelyevv) [reply examples/torch/classification/main.py:269]: The type hint for `train_epoch_fn` defined in the base `TrainingRunner` class is:
```
train_epoch_fn: Callable[[CompressionAlgorithmController, TModel,
                                        Optional[OptimizerType],
                                        Optional[LRSchedulerType],
                                        Optional[int]], None]
```

This didn't match signature of `train_epoch_fn` defined here and in other similar places. So I changed these definitions according to the type hint above (have just found another one for TF object detection which is not fixed yet, will do). 

Regarding the OTX, we are free to decorate train function as we desire so I see no conflict on that part.

Comment (nikita-savelyevv) [reply examples/torch/classification/main.py:269]: Regarding the bad style, one option is to replace unused arguments with `_`. On the other hand it would be less clear for the user in my opinion.

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 4 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:71]: It was decided to improve code quality within a separate task

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 5 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:183]: It was decided to improve code quality within a separate task

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 6 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:254]: It was decided to improve code quality within a separate task

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 8 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:363]: It was decided to improve code quality within a separate task

File: nncf/common/accuracy_aware_training/training_loop.py (hunk 9 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:374]: It was decided to improve code quality within a separate task

File: nncf/tensorflow/accuracy_aware_training/runner.py (hunk 11 above)
Comment (nikita-savelyevv) [reply nncf/tensorflow/accuracy_aware_training/runner.py:111]: It was decided to improve code quality within a separate task

File: nncf/common/accuracy_aware_training/runner.py (hunk 14 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:336]: It was decided to improve code quality within a separate task

File: nncf/tensorflow/accuracy_aware_training/runner.py (hunk 11 above)
Comment (nikita-savelyevv) [reply nncf/tensorflow/accuracy_aware_training/runner.py]: It was decided to improve code quality within a separate task

File: nncf/common/accuracy_aware_training/runner.py [hunk 16]
```diff
@@ -262,6 +313,79 @@ def __init__(self, accuracy_aware_params: Dict[str, object], verbose=True,
...
+        self.optimizer = None
+        self.lr_scheduler = None
+
+        self.base_lr_reduction_factor_during_search = 1.0
+        self.lr_updates_needed = lr_updates_needed
+        self.load_checkpoint_fn = None
+        self.early_stopping_fn = None
+        self.update_learning_rate_fn = None
+
+        self._train_epoch_fn = None
```
Comment (alexsu52) [inline nncf/common/accuracy_aware_training/runner.py:327]: It looks inconsistent when some attributes are prefixed with `_` and some are not.

**Review by vshampor — APPROVED** [review]

File: examples/tensorflow/classification/main.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply examples/tensorflow/classification/main.py:259]: Thanks for you reply. Will TensorFlow classification example dump statistics in the tensorboard after these changes?

File: examples/torch/classification/main.py (hunk 12 above)
Comment (alexsu52) [reply examples/torch/classification/main.py:269]: Essentially, We need to understand where the typo was. If there is the typo in the type hint I will prefer to change type hint.

File: nncf/common/accuracy_aware_training/runner.py (hunk 16 above)
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:207]: Renamed all *fns to start with `_`

File: nncf/common/accuracy_aware_training/runner.py [hunk 17]
```diff
@@ -226,30 +275,32 @@ def create_training_loop(self) -> TrainingRunner:
...
     initialized with the default parameters unless specified in the config.
     """
 
-    def __init__(self, accuracy_aware_params: Dict[str, object], verbose=True,
-                 dump_checkpoints=True):
-        self.maximal_relative_accuracy_drop = accuracy_aware_params.get('maximal_relative_accuracy_degradation')
-        self.maximal_absolute_accuracy_drop = accuracy_aware_params.get('maximal_absolute_accuracy_degradation')
-        self.maximal_total_epochs = accuracy_aware_params.get('maximal_total_epochs', AA_MAXIMAL_TOTAL_EPOCHS)
+    def __init__(self, accuracy_aware_training_params: Dict[str, object], verbose=True,
+                 dump_checkpoints=True, lr_updates_needed=True, **kwargs):
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Removed

File: nncf/common/accuracy_aware_training/runner.py [hunk 18]
```diff
@@ -151,7 +179,27 @@ def load_best_checkpoint(self, model: TModel) -> None:
...
+        """
+        Save a model to the disk.
+
+        :param model: The model to be saved
+        :param checkpoint_path: The path to save the checkpoint to
+        :param compression_controller: The compression controller to be used during
+        model fine-tuning
+        """
+
+    def _load_checkpoint(self, model, checkpoint_path):
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Done

File: nncf/common/accuracy_aware_training/runner.py [hunk 19]
```diff
@@ -151,7 +179,27 @@ def load_best_checkpoint(self, model: TModel) -> None:
         """
         Load the most accurate model state from the fine-tuning history.
 
+        param model: The model object in which the state will be loaded.
+        """
+
+    @abstractmethod
+    def _save_checkpoint(self, model: TModel, compression_controller: CompressionAlgorithmController,
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Done

File: nncf/common/accuracy_aware_training/runner.py [hunk 20]
```diff
@@ -151,7 +179,27 @@ def load_best_checkpoint(self, model: TModel) -> None:
...
+                         checkpoint_path: str) -> None:
+        """
+        Save a model to the disk.
+
+        :param model: The model to be saved
+        :param checkpoint_path: The path to save the checkpoint to
+        :param compression_controller: The compression controller to be used during
+        model fine-tuning
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Done

File: nncf/common/accuracy_aware_training/runner.py [hunk 21]
```diff
@@ -11,25 +11,52 @@
...
+    from torch.utils.tensorboard import SummaryWriter
+
+    TENSORBOARD_AVAILABLE = True
+except ImportError:
+    TENSORBOARD_AVAILABLE = False
+
+try:
+    import matplotlib.pyplot as plt
+    import PIL.Image
+    from torchvision.transforms import ToTensor
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: You are right, fixed this part

File: nncf/common/accuracy_aware_training/runner.py [hunk 22]
```diff
@@ -277,42 +401,135 @@ def initialize_training_loop_fns(self, train_epoch_fn: Callable[[CompressionAlgo
...
+
+        if IMG_PACKAGES_AVAILABLE:
+            plt.figure()
+            plt.plot(self.compressed_training_history.keys(),
+                     self.compressed_training_history.values())
+            buf = io.BytesIO()
+            plt.savefig(buf, format='jpeg')
+            buf.seek(0)
+            image = PIL.Image.open(buf)
+            image = ToTensor()(image)
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: This will cause an error on TF, fixed this part

File: nncf/common/accuracy_aware_training/runner.py [hunk 23]
```diff
@@ -11,25 +11,52 @@
...
 from nncf.config.schemata.defaults import AA_MAXIMAL_TOTAL_EPOCHS
 from nncf.config.schemata.defaults import AA_MINIMAL_COMPRESSION_RATE_STEP
 from nncf.config.schemata.defaults import AA_PATIENCE_EPOCHS
 
+if TYPE_CHECKING:
+    from nncf.torch.accuracy_aware_training.runner import PTAdaptiveCompressionLevelTrainingRunner
+    from nncf.tensorflow.accuracy_aware_training.runner import TFAdaptiveCompressionLevelTrainingRunner
+
+try:
+    from torch.utils.tensorboard import SummaryWriter
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: You are right, this will cause an error on TF, fixed this part

File: nncf/common/accuracy_aware_training/runner.py [hunk 24]
```diff
@@ -214,7 +262,8 @@ def __init__(self, accuracy_aware_training_params: Dict[str, object],
         self.maximal_compression_rate = maximal_compression_rate
         self.dump_checkpoints = dump_checkpoints
 
-    def create_training_loop(self) -> TrainingRunner:
+    def create_training_loop(self) -> Union[PTAdaptiveCompressionLevelTrainingRunner,
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Changed to `BaseAccuracyAwareTrainingRunner`

File: nncf/common/accuracy_aware_training/runner.py [hunk 25]
```diff
@@ -226,30 +275,32 @@ def create_training_loop(self) -> TrainingRunner:
             from nncf.torch.accuracy_aware_training.runner import PTAdaptiveCompressionLevelTrainingRunner #pylint: disable=cyclic-import
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Done

File: examples/torch/classification/main.py (hunk 12 above)
Comment (nikita-savelyevv) [reply examples/torch/classification/main.py]: Changed type hint

File: examples/tensorflow/classification/main.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply examples/tensorflow/classification/main.py]: You are right, it won't, fixed this

**Comment by aiarkinx** [comment]
run openvino pre-commit tests

**Comment by aiarkinx** [comment]
run onnx pre-commit tests

File: nncf/common/accuracy_aware_training/runner.py [hunk 26]
```diff
@@ -151,133 +165,81 @@ def load_best_checkpoint(self, model: TModel) -> None:
...
     initialized with the default parameters unless specified in the config.
     """
 
-    def __init__(self, accuracy_aware_params: Dict[str, object], verbose=True,
-                 dump_checkpoints=True):
-        self.maximal_relative_accuracy_drop = accuracy_aware_params.get('maximal_relative_accuracy_degradation')
-        self.maximal_absolute_accuracy_drop = accuracy_aware_params.get('maximal_absolute_accuracy_degradation')
-        self.maximal_total_epochs = accuracy_aware_params.get('maximal_total_epochs', AA_MAXIMAL_TOTAL_EPOCHS)
+    def __init__(self, accuracy_aware_training_params: Dict[str, object], verbose=True,
+                 dump_checkpoints=True, lr_updates_needed=True):
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:179]: https://github.com/nikita-savelyevv/nncf/blob/actl-improvements/nncf/torch/accuracy_aware_training/runner.py#L79

File: nncf/common/accuracy_aware_training/runner.py [hunk 27]
```diff
@@ -286,33 +248,196 @@ def calculate_minimal_tolerable_accuracy(self, uncompressed_model_accuracy: floa
...
+        if len(possible_checkpoint_rates) == 0:
+            nncf_logger.warning('Could not produce a compressed model satisfying the set accuracy '
+                                'degradation criterion during training. Increasing the number of training '
+                                'epochs')
+            return
+
+        best_checkpoint_compression_rate = max(possible_checkpoint_rates)
+        if best_checkpoint_compression_rate not in self._best_checkpoints:
+            # The checkpoint wasn't saved because it was not fully compressed
+            return
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:407]: I wanted to change the model saving logic to also save the not FULLY_COMPRESSED models (as it is [done ](https://github.com/nikita-savelyevv/nncf/blob/actl-improvements/nncf/common/accuracy_aware_training/runner.py#L253)now), but decided to postpone it until a separate refactoring. 

For now, I added the warning the check for other models as you have suggested.

File: nncf/common/accuracy_aware_training/runner.py [hunk 28]
```diff
@@ -35,13 +39,22 @@
...
+    IMG_PACKAGES_AVAILABLE = False
+
 
 class TrainingRunner(ABC):
     """
     Runner is an object that is used by a TrainingLoop instance to control the training process
     via wrapping user-supplied functions such as `train_epoch_fn` and `validate_fn`.
     """
 
+    CHECKPOINT_PATH_EXTENSION: str
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:57]: Can't figure out how to do it in a good way according to the exact methods you've suggested. But I've made something similar which does the job.

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 29]
```diff
@@ -53,39 +54,26 @@ def run(self, model: TModel, train_epoch_fn, validate_fn, configure_optimizers_f
...
         """
 
 
-class EarlyExitCompressionTrainingLoop(TrainingLoop):
-    """
-    Adaptive compression training loop allows an accuracy-aware training process
-    to reach the maximal accuracy drop
-    (the maximal allowed accuracy degradation criterion is satisfied).
-    """
+class BaseEarlyExitCompressionTrainingLoop(TrainingLoop, ABC):
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:66]: Done

File: nncf/common/accuracy_aware_training/runner.py [hunk 30]
```diff
@@ -286,33 +248,196 @@ def calculate_minimal_tolerable_accuracy(self, uncompressed_model_accuracy: floa
...
+    def _load_checkpoint(self, model: TModel, checkpoint_path: str) -> None:
+        """
+        Load model from path.
+
+        :param model: The model object in which the state will be loaded.
+        :param checkpoint_path: The path where model checkpoint is stored.
+        """
+
+
+class BaseAdaptiveCompressionLevelTrainingRunner(BaseAccuracyAwareTrainingRunner, ABC):
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:347]: Done

File: examples/tensorflow/object_detection/main.py [hunk 31]
```diff
@@ -352,7 +352,7 @@ def run(config):
             timer = Timer()
             timer.tic()
 
-            def train_epoch_fn(compression_ctrl, model, epoch, **kwargs):
+            def train_epoch_fn(compression_ctrl, model, epoch, optimizer, lr_scheduler):
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply examples/tensorflow/object_detection/main.py:355]: Sure. Done.

File: nncf/common/accuracy_aware_training/runner.py [hunk 32]
```diff
@@ -286,33 +248,196 @@ def calculate_minimal_tolerable_accuracy(self, uncompressed_model_accuracy: floa
...
+
+        if is_best_checkpoint:
+            self._save_best_checkpoint(model, compression_controller)
+
+    def configure_optimizers(self):
+        self.optimizer, self.lr_scheduler = self._configure_optimizers_fn()
 
-class BaseAdaptiveCompressionLevelTrainingRunner(BaseAccuracyAwareTrainingRunner):
+    def initialize_training_loop_fns(self, train_epoch_fn, validate_fn, configure_optimizers_fn, dump_checkpoint_fn,
+                                     tensorboard_writer=None, log_dir=None,
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Fixed

**Comment by nikita-savelyevv** [comment]
run pylint tests

**Comment by nikita-savelyevv** [comment]
run pylint test

**Comment by nikita-savelyevv** [comment]
run pylint pre-commit tests

**Comment by nikita-savelyevv** [comment]
run pylint pre-commit test

**Comment by nikita-savelyevv** [comment]
run build

File: nncf/common/accuracy_aware_training/runner.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/common/accuracy_aware_training/runner.py]: > Changed to `BaseAccuracyAwareTrainingRunner`

At least this change is not consistent with
https://github.com/openvinotoolkit/nncf/blob/174ab2667200d3efc376cf3bd0c922f7ed0b1223/nncf/common/accuracy_aware_training/runner_factory.py#L49

[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py]: Changed there also

File: nncf/common/accuracy_aware_training/runner.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:420]: There is one already: 96388

File: nncf/common/accuracy_aware_training/runner.py [hunk 33]
```diff
@@ -142,101 +153,24 @@ def initialize_training_loop_fns(self, train_epoch_fn: Callable[[CompressionAlgo
         :param configure_optimizers_fn: a method to instantiate an optimizer and a learning
         rate scheduler (to be called inside the `configure_optimizers` of the TrainingRunner).
         :param dump_checkpoint_fn: a method to dump a checkpoint.
-        :param tensorboard_writer: The tensorboard object to be used for logging.
-        :param log_dir: The path to be used for logging and checkpoint saving.
         """
 
-    @abstractmethod
-    def load_best_checkpoint(self, model: TModel) -> None:
+    def initialize_logging(self, log_dir: Optional[Union[str, pathlib.Path]] = None,
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/runner.py:158]: Right, done

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 34]
```diff
@@ -125,27 +115,31 @@ def _run_early_exit_training_loop(self, model):
...
-    def print_accuracy_statistics(uncompressed_model_accuracy,
-                                  compressed_model_accuracy,
+    def print_accuracy_statistics(compressed_model_accuracy,
+                                  uncompressed_model_accuracy,
                                   accuracy_drop,
                                   rel_accuracy_drop,
                                   accuracy_budget):
-        nncf_logger.info('Original model accuracy: {:.4f}'.format(uncompressed_model_accuracy))
         nncf_logger.info('Compressed model accuracy: {:.4f}'.format(compressed_model_accuracy))
+        nncf_logger.info('Original model accuracy: {:.4f}'.format(uncompressed_model_accuracy))
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply nncf/common/accuracy_aware_training/training_loop.py:142]: Hmm. Not sure. Reverted back to the original order.

File: tests/torch/accuracy_aware_training/test_training_loop.py [hunk 35]
```diff
@@ -62,9 +62,9 @@ def create_finetuned_lenet_model_and_dataloader(config, eval_fn, finetuning_step
      'final_compression_rate',
      'reference_final_metric'),
     (
-            ({'maximal_relative_accuracy_degradation': 0.01}, 0.66742, 0.996252),
-            ({'maximal_relative_accuracy_degradation': 100.0}, 0.94136, 0.876409),
-            ({'maximal_absolute_accuracy_degradation': 0.10}, 0.767040, 0.938572),
+            ({'maximal_relative_accuracy_degradation': 0.01}, 0.7421, 0.998181),
+            ({'maximal_relative_accuracy_degradation': 100.0}, 0.9165, 0.0),
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply tests/torch/accuracy_aware_training/test_training_loop.py:66]: It means that we don't care about the accuracy of the compressed model, it may be 0. Seems correct, but not something a user would want to do.

[the remark this answers is not in the document]
Comment (alexsu52) [reply tests/torch/accuracy_aware_training/test_training_loop.py:66]: The logic is lame. If test wants to check the maximum compression rate value then change is not correct 0.94136 > 0.9165. Honestly, I don't understand what the test checks.

**Review by alexsu52 — APPROVED** [review]

File: tests/torch/accuracy_aware_training/test_training_loop.py (hunk 35 above)
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply tests/torch/accuracy_aware_training/test_training_loop.py:66]: Yes, will need to revisit existing tests when implementing the ticket 96385.

**Review by kshpv — APPROVED** [review]
Nice improvements!

File: nncf/common/accuracy_aware_training/runner.py [hunk 36]
```diff
@@ -11,21 +11,25 @@
...
 """
 
-from typing import Callable
-from typing import Dict
-from typing import List
-from typing import Optional
-from typing import Union
-from typing import Tuple
-from typing import TypeVar
+from __future__ import annotations
```
Comment (andrey-churkin) [inline nncf/common/accuracy_aware_training/runner.py:14]: Why this import is required here?

**Review by andrey-churkin — APPROVED** [review]

**Comment by andrey-churkin** [comment]
@daniil-lyakhov is absent.
Feel free to left comments and they will be fixed in follow-up PRs.
